pkg/operator/controller/garden/reconciler_delete.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package garden
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpa"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+func (r *Reconciler) delete(
+	ctx context.Context,
+	log logr.Logger,
+	garden *operatorv1alpha1.Garden,
+	secretsManager secretsmanager.Interface,
+) (
+	reconcile.Result,
+	error,
+) {
+	log.Info("Destroying gardener-resource-manager")
```

Comment:
We are not doing this in the `gardenlet'`s `Seed` controller. 👀 Why do you mention it here? Probably we should follow the same patterns in both controllers. Anyways, do you suggest to first delete all `ManagedResource`s before destroying `gardener-resource-manager`?

Reply:
Yes, I saw that we are not doing this for Seeds either. 
I suggest only waiting for all `ManagedResources` to be gone, but not deleting them (both in garden and seed controller).

There might be `ManagedResources` left over from other components (not managed by gardenlet/gardener-operator). In this case, it would be a better user experience to block deletion with a proper error message and let the (human) operator cleanup manually. However, as long as resource-manager is still there, cleaning up leftovers is much simpler.

Reply:
OK, understood, I can add the check in both controllers!